### ISSUE TITLE
Ambiguity with "redeclare" as an element (as opposed to a modification).

### ISSUE THREAD

**Issue by modelica-trac-importer** [modelica/ModelicaSpecification#468]
**Reported by stefanv on 8 Dec 2010 19:17 UTC**
Section 7.3 of the Modelica 3.2 specification states:

_A redeclare construct as an element replaces the inherited declaration of a local class or component with another declaration._

This is unclear when the redeclared entity is used in the class from which it was inherited. For example, consider:

```
model A
    replaceable package P
        constant Real x = 1;
    end P;
    constant Real y = P.x;
end A

model B
    extends A;
    redeclare package P constant Real x = 2; end P;
    constant Real z = P.x;
end B;
```

It is clear that the value of B.z should be 2, since the redeclaration has replaced the P inherited from A with the P defined in B.

However, it is not clear whether the value of B.y should be 1 or 2.

Another way of phrasing the same ambiguity is to ask whether the following is equivalent to the code above:

```
model A
    replaceable package P
        constant Real x = 1;
    end P;
    constant Real y = P.x;
end A

model B
    extends A(redeclare package P constant Real x = 2; end P);
    constant Real z = P.x;
end B;
```

In this case, it is clear that both B.y and B.z have the value 2.

In other words, is having a redeclaration appearing as an element equivalent to having it appear as a modification in all the extends-clauses that extend from classes containing the entity to be redeclared?

----

Migrated-From: https://trac.modelica.org/Modelica/ticket/468

**Comment by modelica-trac-importer** [comment]
**Comment by kurzbach on 8 Mar 2011 13:39 UTC**
Beside of the sentences in 7.3 it is unclear and unspecified how redeclare elements are 
applied during the flattening process (chapter 5.3). This needs to be changed.

Here is the example from your email, slightly extended, which shows the behavior of the current implementation in Dymola:

```
model M
	model A
		replaceable model AT
			Real x;
			equation
			x = 1;
		end AT;
		model BT
			extends AT;
		end BT;
		AT at;
	end A;

	extends A;

	redeclare model AT
		Real x;
		equation
		x = 2;
	end AT;

	AT bt;
	BT bt2;
	equation
		assert(at.x==1,"Error: at.x<>1");
		assert(bt.x==2,"Error: bt.x<>2");
		assert(bt2.x==2,"Error: bt2.x<>2");
end M;

```

The asserts show the results and it is unclear from the specification why `at.x` <> `bt2.x`. 
Why is the `redeclare AT` applied in `A` to `extends AT;` but not to `AT at;` ?

**Comment by modelica-trac-importer** [comment]
**Modified by kurzbach on 23 Mar 2011 16:29 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 23 Mar 2011 17:39 UTC**
Second example is illegal since it extends a replaceable class, however, this was not the main point of the example.

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 23 Mar 2011 18:00 UTC**
The Modelica.Media package (saturationPressure in PartialTwoPhaseMedium and use in setState_T) seems to assumes the interpretation where P.y is 2.

One possibility would be to clarify that "A redeclare construct as an element replaces the inherited declaration of a local class or component with another declaration." 

to say that

"A redeclare construct as an element replaces the inherited declaration of a local class or component with another declaration; this new declaration is used also in the base-classes." 

However, Gerd Kurzbach says other semantics is used and depends on the ordering of the flattening process in section 5.6.

Conclusion: more discussion and clarification is needed.

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 24 Mar 2011 17:34 UTC**
(Did not find any such examples in use.)

Proposed changes:
1. Section 7.3 should have no difference between modifier-redeclare and element-redeclare (the 2nd sentence was written differently since element-redeclare always apply to inherited elements - have that as separate sentence).
Favor: 10 Against: 0

2. Prohibit modifier-redeclare and element-redeclare in the same lexical scope of the same element, i.e.:
```
  extends A(redeclare model B=...);
  redeclare model B=...;
```
Favor: 9 Against: 1

Agreed.

3. In 5.6.1 clarify that 'redeclare take effect' - include redeclared elements, see section 7.3. A bit unclear how.

**Comment by modelica-trac-importer** [comment]
**Comment by HansOlsson on 24 Mar 2011 17:44 UTC**
In 5.6.1 the problem might be seen as where 'element redeclare' are applied. One possibility would to be to add the element redeclare to the environment - still not clear on how.

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 17 May 2011 11:25 UTC**

How about class extends?
Does redeclare class extends replace ONLY the declaration 
inherited in the local scope or does it apply to base classes 
also? 
Dymola seems to only apply the redeclare class extends 
to the declarations inherited in the local scope.
Example:
```
package ClassExtendsApplication

  class A
    Real a;
  end A;
  
  class B
    constant Real x = 2;
    
    replaceable class C
      extends A;
      parameter Real y = x;
    end C;

    C bc;
  end B;
  
  class D
    extends B(x = 4);
    
    redeclare class extends C(y = 5)
      Real newComp(start = y);
    end C;
    
    C dc;
  end D;

end ClassExtendsApplication;
```

So the question is what does the instantiation of ClassExtendsApplication.D gives you? Dymola gives:
```
model ClassExtendsApplication.D
constant Real x = 4;
parameter Real bc.y = 4;
parameter Real dc.y = 5;
Real bc.a;
Real dc.a;
Real dc.newComp(start = dc.y);
// No equations
end ClassExtendsApplication.D;
```

Is this correct? 
Shouldn't the "redeclare class extends C(y = 5) ..."
be also applied to the base class "B", i.e.:
```
class ClassExtendsApplication.D
  constant Real x = 4.0;
  Real bc.a;
  parameter Real bc.y = 5.0;
  Real bc.newComp(start = bc.y);
  Real dc.a;
  parameter Real dc.y = 5.0;
  Real dc.newComp(start = dc.y);
end ClassExtendsApplication.D;
```

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Comment by adrpo on 20 May 2011 15:57 UTC**
Yet another issue with this. In the specification we have:
```
7.3 Redeclaration
A modifier with the keyword replaceable is automatically seen as being a redeclare.
```

And then we have an example (don't mind that Dymola doesn't parse it, is correct):
```
model HeatExchanger
  replaceable parameter GeometryRecord geometry;
  replaceable input Real u[2];
end HeatExchanger;

HeatExchanger(
 /* redeclare */ replaceable /* parameter */ GeoHorizontal geometry,
 redeclare /* input */ Modelica.SIunits.Angle u /*[2]*/);
 // The semantics ensure that parts in /*.*/ are automatically added
 // from (the original declaration <-- i added this missing bit)
``` 

Now how about this model:
```
model X
 extends HeatExchanger;
 replaceable GeoHorizontal geometry; // <- do we add redeclare to this?
end X;
```
Do we add redeclare to the replaceable element if there is
no difference between redeclare as element and redeclare with 
modifier?
If we don't add the redeclare we might get duplicate elements with
different type. If we add the redeclare we might actually not have
a replaceable element in the base class.

So we should pick one of:
- add the redeclare automatically (this might change the 
  semantics of some models) as we do for modifiers or 
- specify that element replaceable \*does not automatically 
  means redeclare\* if redeclare is not explicitly given.
- require redeclare to be given explicit also for modifiers

Cheers,
Adrian Pop/

**Comment by modelica-trac-importer** [comment]
**Modified by hansolsson on 28 Sep 2011 12:46 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by stefanv on 28 Sep 2011 13:24 UTC**
What was the agreed "fix"?

**Comment by modelica-trac-importer** [comment]
**Modified by stefanv on 28 Sep 2011 19:46 UTC**

**Comment by modelica-trac-importer** [comment]
**Comment by hansolsson on 29 Sep 2011 15:31 UTC**
The fix is now on the svn-server.

As for the latest addition above: it is a separate issue 'Should redeclare be required for redeclare-modifiers', and it is unclear whether it is question about semantics (if so the answer is the 2nd item in the list) or a proposal for change. When such a new issue is found, please please open a new ticket for that issue, and do not add it on an existing ticket. Even if the tickets are related it just creates a mess, and thus I close this one.

Basically the intent of the comment above 'Section 7.3 should have no difference...' was that the different form of redeclare should work in the same way - and that restrictions only applying to one of them should be listed separately so that the same behaviour is not specified differently, since that causes confusion.

However, even if they behave the same way that are not identical (at least not syntactically), as was clear from the addition of the restriction that modifiers can only use short class definitions (#521). That would have been a major incompatible change if it also applied to element-redeclaration; and the 'redeclare' on modifiers is a similar form of syntactic restriction.
